src/coreclr/jit/optimizer.cpp

```diff
@@ -3546,8 +3569,14 @@ bool Compiler::optComputeLoopRep(int        constInit,
 // Loops handled are fully unrolled; there is no partial unrolling.
 //
 // Limitations: only the following loop types are handled:
-// 1. "while" loops
-// 2. constant bound loops
+// 1. "while" loops (top entry)
+// 2. constant initializer, constant bound
+// 3. The entire loop must be in the same EH region.
+// 4. The loop iteration variable can't be address exposed.
+// 5. The loop iteration variable can't be a promoted struct field.
+// 6. We must be able to calculate the total constant iteration count.
+// 7. On x86, there is a limit to the number of return blocks. So if there are return blocks in the loop that
```

Comment:
Hopefully by now we've moved any such blocks out of the loop range. Can you look into whether this is still a true constraint?

Reply:
As far as I can tell, we have moved all BBJ_RETURN out of the loop body. I can't see any case where we haven't, but I don't know how to prove that currently. I could follow-up by adding an assert to the new loop checker that there are no BBJ_RETURN blocks.

Reply:
Actually, it looks like there are cases where we don't move BBJ_RETURN out of the loop body. I don't know the precise conditions. SPMI found coreclr test `Program:M49()` and `System.Security.Cryptography.Pkcs.Rfc3161TimestampToken:TryGetCertIds()` (which has lots of nested loops + return + EH).

This test creates a case that fails:
```
        for (int i = 0; i < Vector<int>.Count; i++)
        {
            a += 1;
            if (a > 12000) return a;
            try {
                a += 2;
            } finally {
                a += 3;
            }
        }
```

The block table is:
```
-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1       [000..006)-> BB06 ( cond )                     i
BB02 [0001]  2       BB01,BB05             0.50    [006..012)-> BB04 ( cond )                     i bwd bwd-target
BB03 [0002]  1       BB02                  0.50    [012..014)        (return)                     i
BB04 [0004]  1  0    BB02                  0.50    [014..01B)                 T0      try { }     keep i try bwd
BB05 [0011]  1       BB04                  0.50    [01B..02C)-> BB02 ( cond )                     keep i bwd cfb cfe
BB06 [0008]  2       BB01,BB05             0.50    [02C..02E)        (return)                     i
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ funclets follow
BB07 [0005]  1     0                       0       [01B..020)        (finret)    H0 F fault { }   keep i rare flet bwd
-----------------------------------------------------------------------------------------------------------------------------------------
```
and the loop is from BB02..BB05 with an exit at BB03.

Reply:
Interestingly, presuming we ever would unroll with an embedded return, we check to avoid doing so on x86 if it would exceed the max allowed epilog limit, but not elsewhere. Elsewhere, we still limit return blocks to 4 in fgAddInternal, but it's not a hard limit, so presumably we could exceed it in unrolling.